### ISSUE TITLE
Derivatives don't need their own memory

### ISSUE THREAD

**Issue by kparichay** [nnstreamer/nntrainer#765]
The output of a layer can be used to store the derivative from the next layer.
The reason for this is that the output of a layer is not used in the backwarding (except for softmax backwarding which can be handled as a special). As the shape of derivative and output layer is exactly same, this saves the memory of all the layer return derivatives, reducing the peak memory by a LOT.

With training, the memory impact is usually doubled with gradients and return derivatives. 
With this scheme, the memory impact from return derivatives can be completely eliminated, bringing the memory requirement for training very close to the memory requirement for inference.
